lumicks/pylake/population/mixture.py

```diff
@@ -0,0 +1,178 @@
+import numpy as np
+from sklearn.mixture import GaussianMixture
+from scipy import stats
+
+
+def as_sorted(fcn):
+    """Decorator to return results sorted according to mapping array."""
+
+    def wrapper(self, *args, **kwargs):
+        result = fcn(self, *args, **kwargs)
+        return result[self._map]
+
+    return wrapper
+
+
+class GaussianMixtureModel:
+    def __init__(self, trace, K, init_method, n_init, tol, max_iter):
+        """A wrapper around scikit-learn's GMM
+
+        Parameters
+        ----------
+        trace : Slice-like
+            data object used for model training
+        K : int
+            number of Gaussian components in the model
+        init_method : 'kmeans' or 'random'
+            method used to initialize parameters
+        n_init : int
+            number of initializations to perform
+        tol : float
+            convergence tolerance
+        max_iter : int
+            maximum number of iterations to perform
+        """
+        self.K = K
+        self._training_data = trace
```

Comment:
Do think about if you really need to store a full `Slice` as the training data. It's a tradeoff:
* The downside is that users will need to pass the slice again for the plotting function.
* The upside is that you don't need to store and think about all the extra stuff in `Slice` that you don't need for the training. It removes the coupling between the GMM and `Slice`. And it makes your testing easier because you no longer need a full `Slice` for unit testing.

Reply:
I kept going back and forth on this (see Joep's earlier comments). Looking at it again, I think I agree that it's not worth it to store the `Slice` or training data inside the object.

One thing that I still find awkward here though is now the constructor accepts a `np.array` as training data but other methods (`.plot()`, `.hist()`, `.label()`, etc...) rely on the input argument being `Slice`-like (ie, having a `.data` attribute and sometimes a `.plot()` method).

In addition to force channels, I could see this class being useful also for `PointScan.color_photon_count` which returns `Continuous` or output from `KymoLine.sample_from_image()` which is currently returns a `list`. For handling these situations, I don't know what the best option would be:

- enforce the `Slice`-like API, but `Continuous.plot()` does not exist and `KymoLine.sample_from_image()` would need to return a `Slice` instead of `list`
- have all of the `GMM` methods just accept `np.array`, although I think `GMM.plot(trace.data)` is still a bit awkward
- have some sort of internal check in these methods to handle the different types that could be passed as input arguments

What do you guys think?

Reply:
The photon counts are returned as `Slice`s as far as I know. Nothing should return `Continuous` since that's an implementation detail and not a public API. If something does return that, it should be fixed.

I think that it's fine to require `Slice` as the input. The plotting functions don't just use the data but also timestamps, seconds, and labels as far as I can see. I do think that that the core processing should be done with raw data, but I don't worry so much about the plotting functions. They are more of a convenience feature and it's hard to make something both convenient and also very generic. If someone needs to plot the raw data, it's easy to see the source code and replicate it.

That said, I also don't like the discrepancy between the core class and the plotting functions. If this was a different language, I would say you should make the plotting functions into free functions that take the model as one of the arguments. But that's difficult to do in Python -- or just as awkward as the current situation. So I would say that this is fine for now and we'll discover more as we get info about how this API will actually be used. Too much guessing for now.

Reply:
you're right, photon counts is a `Slice`, it's just a very convoluted chain of calls and inheritance so I got confused